### ISSUE TITLE
Map Matching API not matching motorways in a local instance

### ISSUE THREAD

**Issue by pmgomesLoL** [valhalla/valhalla#4886]
### Discussed in https://github.com/valhalla/valhalla/discussions/4884

<div type='discussions-op-text'>

<sup>Originally posted by **pmgomesLoL** September  9, 2024</sup>
I've setup a local server (docker) and as a teste i've tryed the following request. If i send it to my server none of the motorways are matched, if i sendit to the oficial "demo/teste" server the route is matched all the way

Thanks for you help

Best regards
Pedro 

```
{
    "shape": [
        {
            "lat": 40.952171,
            "lon": -8.635384,
            "heading": 65,
            "type": "break",
            "time": 1725605220
        },
        {
            "lat": 40.952094,
            "lon": -8.635365,
            "heading": 65,
            "type": "through",
            "time": 1725605238
        },
        {
            "lat": 40.952252,
            "lon": -8.635282,
            "heading": 202,
            "type": "through",
            "time": 1725605407
        },
        {
            "lat": 40.952252,
            "lon": -8.635282,
            "heading": 202,
            "type": "through",
            "time": 1725605586
        },
        {
            "lat": 40.952252,
            "lon": -8.635282,
            "heading": 202,
            "type": "through",
            "time": 1725605765
        },
        {
            "lat": 40.952252,
            "lon": -8.635282,
            "heading": 202,
            "type": "through",
            "time": 1725605945
        },
        {
            "lat": 40.953129,
            "lon": -8.636815,
            "heading": 297,
            "type": "through",
            "time": 1725606124
        },
        {
            "lat": 40.960919,
            "lon": -8.629563,
            "heading": 357,
            "type": "through",
            "time": 1725606304
        },
        {
            "lat": 40.964282,
            "lon": -8.629965,
            "heading": 183,
            "type": "through",
            "time": 1725606483
        },
        {
            "lat": 40.965471,
            "lon": -8.619191,
            "heading": 64,
            "type": "through",
            "time": 1725606663
        },
        {
            "lat": 40.958505,
            "lon": -8.609175,
            "heading": 171,
            "type": "through",
            "time": 1725606842
        },
        {
            "lat": 40.909855,
            "lon": -8.60684,
            "heading": 160,
            "type": "through",
            "time": 1725607022
        },
        {
            "lat": 40.859305,
            "lon": -8.570744,
            "heading": 160,
            "type": "through",
            "time": 1725607201
        },
        {
            "lat": 40.802535,
            "lon": -8.557593,
            "heading": 191,
            "type": "through",
            "time": 1725607381
        },
        {
            "lat": 40.784044,
            "lon": -8.559904,
            "heading": 175,
            "type": "through",
            "time": 1725607470
        },
        {
            "lat": 40.783949,
            "lon": -8.559926,
            "heading": 191,
            "type": "through",
            "time": 1725607488
        },
        {
            "lat": 40.782864,
            "lon": -8.548799,
            "heading": 118,
            "type": "through",
            "time": 1725607560
        },
        {
            "lat": 40.735176,
            "lon": -8.523914,
            "heading": 166,
            "type": "through",
            "time": 1725607740
        },
        {
            "lat": 40.677963,
            "lon": -8.519559,
            "heading": 187,
            "type": "through",
            "time": 1725607919
        },
        {
            "lat": 40.623795,
            "lon": -8.536952,
            "heading": 200,
            "type": "through",
            "time": 1725608098
        },
        {
            "lat": 40.573571,
            "lon": -8.561194,
            "heading": 165,
            "type": "through",
            "time": 1725608278
        },
        {
            "lat": 40.517753,
            "lon": -8.547299,
            "heading": 138,
            "type": "through",
            "time": 1725608457
        },
        {
            "lat": 40.467862,
            "lon": -8.521366,
            "heading": 171,
            "type": "through",
            "time": 1725608637
        },
        {
            "lat": 40.418938,
            "lon": -8.496644,
            "heading": 167,
            "type": "through",
            "time": 1725608816
        },
        {
            "lat": 40.366164,
            "lon": -8.49294,
            "heading": 172,
            "type": "through",
            "time": 1725608996
        },
        {
            "lat": 40.310175,
            "lon": -8.489198,
            "heading": 169,
            "type": "through",
            "time": 1725609175
        },
        {
            "lat": 40.271478,
            "lon": -8.476302,
            "heading": 110,
            "type": "through",
            "time": 1725609355
        },
        {
            "lat": 40.257442,
            "lon": -8.44342,
            "heading": 175,
            "type": "through",
            "time": 1725609534
        },
        {
            "lat": 40.221424,
            "lon": -8.439684,
            "heading": 167,
            "type": "through",
            "time": 1725609714
        },
        {
            "lat": 40.220181,
            "lon": -8.439018,
            "heading": 166,
            "type": "through",
            "time": 1725609893
        },
        {
            "lat": 40.223713,
            "lon": -8.425256,
            "heading": 69,
            "type": "through",
            "time": 1725610073
        },
        {
            "lat": 40.226014,
            "lon": -8.408801,
            "heading": 106,
            "type": "through",
            "time": 1725610148
        },
        {
            "lat": 40.214507,
            "lon": -8.395362,
            "heading": 204,
            "type": "through",
            "time": 1725610252
        },
        {
            "lat": 40.210003,
            "lon": -8.402526,
            "heading": 269,
            "type": "through",
            "time": 1725610344
        },
        {
            "lat": 40.205148,
            "lon": -8.406357,
            "heading": 186,
            "type": "through",
            "time": 1725610431
        },
        {
            "lat": 40.203975,
            "lon": -8.407931,
            "heading": 103,
            "type": "through",
            "time": 1725610611
        },
        {
            "lat": 40.203975,
            "lon": -8.407931,
            "heading": 103,
            "type": "break",
            "time": 1725610675
        }
    ],
    "costing": "auto",
    "costing_options": {
        "auto": {
            "maneuver_penalty": 5,
            "country_crossing_penalty": 0,
            "country_crossing_cost": 600,
            "width": 1.6,
            "height": 1.9,
            "use_highways": 1,
            "use_tolls": 1,
            "use_ferry": 1,
            "ferry_cost": 300,
            "use_living_streets": 0.5,
            "use_tracks": 0,
            "private_access_penalty": 450,
            "ignore_closures": false,
            "ignore_restrictions": false,
            "ignore_access": false,
            "closure_factor": 9,
            "service_penalty": 15,
            "service_factor": 1,
            "exclude_unpaved": 1,
            "shortest": false,
            "exclude_cash_only_tolls": false,
            "top_speed": 140,
            "fixed_speed": 0,
            "toll_booth_penalty": 0,
            "toll_booth_cost": 15,
            "gate_penalty": 300,
            "gate_cost": 30,
            "include_hov2": false,
            "include_hov3": false,
            "include_hot": false,
            "disable_hierarchy_pruning": false
        }
    },
    "shape_match": "walk_or_snap",
    "trace_options": {
        "gps_accuracy": 10,
        "search_radius": 10,
        "use_timestamps": true,
        "breakage_distance": 200000
    }
}
```

On the backend serve logs I have the follwing output

```
valhalla_latest  | 55 2024/09/09 10:53:44.562475 POST /trace_route HTTP/1.1
valhalla_latest  | 2024/09/09 10:53:44.562926 [INFO] Got Loki Request 55
valhalla_latest  | 2024/09/09 10:53:44.567185 [INFO] Got Thor Request 55
valhalla_latest  | 2024/09/09 10:53:44.572924 [WARN] walk_or_snap algorithm failed to find exact route match; Falling back to map_match...
valhalla_latest  | 2024/09/09 10:53:44.585286 [INFO] Got Odin Request 55
valhalla_latest  | 55 2024/09/09 10:53:44.587209 200 25504
```
</div>

**Comment by ImreSamu** [comment]
The demo/test server was upgraded in the last 24 hours to version >=3.5.0  (  https://valhalla1.openstreetmap.de/status )

IMHO: In your test case, the Valhalla versions were likely different.
* Related issue:  https://github.com/valhalla/valhalla/issues/4894#issuecomment-2359041066  

If this issue is still relevant, could you test it again?

**Comment by ImreSamu** [comment]
@pmgomesLoL :

> I'm inclined to think that the problem I'm having is related to the default configuration.

If I understand correctly, the  Valhalla-**"Update public FOSSGIS servers"**  got stuck, 
and until a few days ago, it was still running a version from about 5 months ago.
* https://github.com/valhalla/valhalla/actions/workflows/update_public_servers.yml

This means that your previous testing was done based on that version.

Could you also specify which Docker image you used for testing?
 (As there are at least three, and several different versions of each.)

That's why it's important to retest and confirm whether this is a regression—meaning an existing feature broke (and the version from 5 months ago was still working fine).
Or, if you've already retested, could you make that clear?

Of course, there could be many other reasons (configuration issues, differences in the administrative database, data problems, or other unknown issues), but these are less likely.

> Is it possible to check if the demo/test server is running any custom configuration and if so, share it?

I can't help with that. I'm just a regular Valhalla user.

**Comment by pmgomesLoL** [comment]
I'm currently building form the source to try and Debug the issue localy but using the latest docker the problem manifest it self.

Also I'm using the latest Portugal map extract at https://download.geofabrik.de/europe/portugal-latest.osm.pbf

Best regards 

Pedro

**Comment by ImreSamu** [comment]
> But I was using the latest docker  ...

If possible, please provide detailed steps on how you conducted the tests. 

The parameters you used are especially important!  https://github.com/gis-ops/docker-valhalla?tab=readme-ov-file#environment-variables 

For example what is your settings for
- `build_elevation`
- `build_admins`
- `build_time_zones`
- ....

> I'm also using the latest Portugal map extract 

note: 
The latest version is not ideal for reproducible local testing, 
either in terms of data or the Docker image, as it changes daily or weekly.

Please try to be more specific, for example:
* data: `https://download.geofabrik.de/europe/portugal-240920.osm.pbf`
   * Check the Geofabrik "raw directory index"
* dockerimage: `docker pull ghcr.io/gis-ops/docker-valhalla/valhalla:latest@sha256:6cd288f95ac0c37df9dc7b7e00c51a4514ae0d901996fea81eb779a85835db94`
   * Check the x86-64 OS/Arch https://github.com/gis-ops/docker-valhalla/pkgs/container/docker-valhalla%2Fvalhalla
   
> but locally it still has the same problem.

Also, provide the exact request you made and both responses you received, from both the local server and the FOSSGIS server.

If you have an extreme simple test program that can test both (local and FOSSGIS servers), that would be even better. 
In other words, try to make it as easy as possible for others to run the tests.

Currently, there are many issues with `trace_route` and `map-matching`.  -->  https://github.com/orgs/valhalla/projects/2
We need a reproducible problem description that will still be valid 1-2-3 months from now, in case the solution isn’t found immediately.

**Comment by ImreSamu** [comment]
> I'm also using the latest Portugal map extract at https://download.geofabrik.de/europe/portugal-latest.osm.pbf

As I see it, part of the problem is that the Portuguese OSM extract doesn't contain the entire country, 
which causes the Portugal administrative boundaries to be incomplete. 
https://download.geofabrik.de/europe/portugal.html
![image](https://github.com/user-attachments/assets/531154f5-33f5-431e-b730-de70a9c6551e)

Some islands - and thus Portugal's maritime boundaries - are missing. 
- [Portugal (águas territoriais)](https://www.openstreetmap.org/way/1135405233#map=11/30.0938/-15.8409)
However, the data is included in the full openstreetmap planet database.

As a result, Valhalla (`build_admins`)  can't generate accurate data from it. 
When I manually fix it, the local Valhalla produces relatively better results, 
but it's still not exactly same as a FOSSGIS server result.

It would be possible to test a local Valhalla planet as well, but that would be quite resource-intensive.

I probably won't be able to help any further with this issue.

**Comment by pmgomesLoL** [comment]
Hi, @ImreSamu

> It would be possible to test a local Valhalla planet as well, but that would be quite resource-intensive.

I'll setup a planet local instance, to check if the problem presists, it will take some time, i'll post the result when it's done

Thanks for the help, 

Pedro

**Comment by nilsnolde** [comment]
closing due to inactivity
